celery/backends/redis.py

```diff
@@ -128,9 +128,9 @@ def reconnect_on_error(self):
         except self._connection_errors:
             try:
                 self._ensure(self._reconnect_pubsub, ())
-            except self._connection_errors:
+            except self._connection_errors as e:
                 logger.critical(E_RETRY_LIMIT_EXCEEDED)
-                raise
+                raise RuntimeError(E_RETRY_LIMIT_EXCEEDED) from e
```

Comment:
@Nusnus would you be able to help me understand if a `RuntimeError` is what we'd want to raise here, or if there's a more specific exception available to us in Python (in general), or in this repository (perhaps there's an error class that we'd want to subclass) that we can raise here so that clients waiting for task results can more easily capture and handle?

Reply:
To add a bit more context -- previously, this error (`e`) was being raised on the green thread, causing it to exit, but that error was never checked/propagated to the main/calling thread.

The change in this PR now propagates any errors raised from `result_consumer.drain_events` on the green thread to the calling thread, so this error would now be exposed to end-users in the event that redis max retries are reached.  

The reason we're opting to throw a new error here is to provide the end-user with the error message:

```
Retry limit exceeded while trying to reconnect to the Celery redis result store backend. 
The Celery application must be restarted.
```

Without this change, we would be throwing a `ConnectionError`, and the message would be lacking the context on the action the user (or app) needs to take to restore service. We're under the impression `ConnectionError` implies the error may be transient and could "fix itself" given time.  

That said, `RuntimeError` feels generic/too broad for the app to catch and enter a graceful shutdown on, but we're not sure how the celery project feels about creating/exporting their own errors for this situation.